### PR TITLE
RCA graph to build a temperature profile of a cluster

### PR DESCRIPTION
**Pull request by yojs**
*Issue #, if available:*
#143

*Description of changes:*
These changes can be best described in this [README](https://github.com/opendistro-for-elasticsearch/performance-analyzer-rca/blob/aa35434ab363be60df24359ed3527af0a9844a71/src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/temperature/README.md)

*Tests:*
There are tests to check that the RCAs are generated.

*Code coverage percentage for this patch:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/Metric.java [hunk 1]
```diff
@@ -48,7 +48,7 @@
 
   public Metric(String name, long evaluationIntervalSeconds) {
     super(0, evaluationIntervalSeconds);
-    this.name = name;
+    this.name = name.isEmpty() ? this.getClass().getSimpleName() : name;
```
[the remark this answers is not in the document]
Comment (yojs) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/Metric.java:51]: Metric cannot be null.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/flow_units/temperature/DimensionalTemperatureFlowUnit.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    public DimensionalTemperatureFlowUnit(long timeStamp, NodeDimensionalTemperatureSummary nodeDimensionProfile) {
+        super(timeStamp, ResourceContext.generic(), nodeDimensionProfile, true);
+        this.nodeDimensionProfile = nodeDimensionProfile;
+    }
+
+    // A dimension flow unit never leaves a node. So, we don't need to generate protobuf messages.
+    @Override
+    public FlowUnitMessage buildFlowUnitMessage(String graphNode, String esNode) {
+        throw new IllegalArgumentException("This should not be called.");
```
[the remark this answers is not in the document]
Comment (yojs) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/flow_units/temperature/DimensionalTemperatureFlowUnit.java:47]: Sounds good. I will change it.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/CompactClusterLevelNodeSummary.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+     * @return Summary object constructed from the database row
+     */
+    public static CompactClusterLevelNodeSummary build(Record record) {
+        String nodeId = record.get(NODE_ID_COL_NAME, String.class);
+        String hostIp = record.get(HOST_IP_ADDRESS_COL_NAME, String.class);
+
+        CompactClusterLevelNodeSummary summary = new CompactClusterLevelNodeSummary(nodeId, hostIp);
+
+        for (TemperatureVector.Dimension dimension : TemperatureVector.Dimension.values()) {
+            Short mean = record.get(dimension.NAME + MEAN_SUFFIX_KEY, Short.class);
```
Comment (rguo-aws) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/CompactClusterLevelNodeSummary.java:52]: do we need to handle DataType exception here ?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/core/temperature/ShardStore.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ * The same set of shards will be seen for each metric and across multiple operations.
+ * This creates a pool of all shards so that they can be referenced from multiple places.
+ */
+public class ShardStore {
+    private static final Logger LOG = LogManager.getLogger(ShardStore.class);
+    /**
+     * The key for the outer map is indexName. The key for inner map is the ShardID. Given an
+     * indexName and shardId, a shard can be uniquely identified.
+     */
+    Map<String, Map<Integer, ShardProfileSummary>> list;
```
Comment (rguo-aws) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/core/temperature/ShardStore.java:37]: Can we reuse Khushboo's IndexShardKey object as the key in this map ? The hashcode() in IndexShardKey is logically equivalent to the hashcode used in ElasticSearch to keep track of shards.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/flow_units/temperature/ClusterTemperatureFlowUnit.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.temperature.ClusterTemperatureSummary;
+
+public class ClusterTemperatureFlowUnit extends ResourceFlowUnit {
+    private final ClusterTemperatureSummary clusterTemperatureSummary;
+
+    public ClusterTemperatureFlowUnit(long timeStamp, ResourceContext context,
+                                      ClusterTemperatureSummary resourceSummary) {
+        super(timeStamp, context, resourceSummary, true);
+        clusterTemperatureSummary = resourceSummary;
+    }
```
[the remark this answers is not in the document]
Comment (yojs) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/flow_units/temperature/ClusterTemperatureFlowUnit.java:29]: Good point ! Added that

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/metric/AggregateMetric.java [hunk 6]
```diff
@@ -44,63 +44,104 @@
...
-  private final AggregateFunction aggregateFunction;
-
-  public AggregateMetric(final long evaluationIntervalSeconds, final String tableName,
-      final AggregateFunction aggregateFunction, final String... groupByFieldsName) {
-    super(AggregateMetric.NAME, evaluationIntervalSeconds);
-    this.tableName = tableName;
-    this.groupByFieldsName = new ArrayList<>(Arrays.asList(groupByFieldsName));
-    this.aggregateFunction = aggregateFunction;
-  }
+    private static final Logger LOG = LogManager.getLogger(AggregateMetric.class);
```
[the remark this answers is not in the document]
Comment (yojs) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/metric/AggregateMetric.java:47]: Fixed that

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/CompactClusterLevelNodeSummary.java (hunk 3 above)
Comment (yojs) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/CompactClusterLevelNodeSummary.java:52]: Good point. I will handle them

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/core/temperature/ShardStore.java (hunk 4 above)
Comment (yojs) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/core/temperature/ShardStore.java:37]: That code is not merged yet. I will add a TODO to be addressed later. Thanks for the pointer.

**Pull request closed without merging**